File: l33tcode/insertion-sort-list.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def insertionSortList(self, head):
        fake_inf = ListNode(float("-inf"))
        node = head
        prev_node = fake_inf

        while node:
            if prev_node.val > node.val:
                inner_node = head
                inner_prev_node = fake_inf

                while inner_node:
                    if inner_node.val > node.val:
                        if inner_node == head:
                            head = node
                        prev_node.next = node.next
                        inner_prev_node.next = node
                        node.next = inner_node
                        node = prev_node.next
                        break
                    inner_prev_node = inner_node
                    inner_node = inner_node.next
            else:
                prev_node = node
                node = node.next

        return head


x = ListNode(4)
x.next = ListNode(2)
x.next.next = ListNode(1)
x.next.next.next = ListNode(3)

s = Solution()
h = s.insertionSortList(x)

print(h.val)
print(h.next.val)
print(h.next.next.val)
print(h.next.next.next.val)

File: l33tcode/insertion-sort-list.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func insertionSortList(head *ListNode) *ListNode {
	fakeHead := &ListNode{} // just a node with no value, we'll never check it
	fakeHead.Next = head

	prevNode := fakeHead
	node := fakeHead.Next

	for node != nil {
		prevNodeInsert := fakeHead
		nodeInsert := fakeHead.Next

		for nodeInsert != nil && nodeInsert.Val < node.Val && nodeInsert != node {
			prevNodeInsert = nodeInsert
			nodeInsert = nodeInsert.Next
		}

		nodeNext := node.Next

		if nodeInsert != node {
			prevNodeInsert.Next = node

			node.Next = nodeInsert

			prevNode.Next = nodeNext
		} else {
			prevNode = node
		}

		node = nodeNext
	}

	return fakeHead.Next
}
